---
layout: page
title: About me
meta-title: Avi Karn, Ph.D.
subtitle: Quantitative & Computational Geneticist
---
<div id="aboutme-section">

<p class="about-text">

<div id="aboutme-section">

<div class="LI-profile-badge"  data-version="v1" data-size="large" data-locale="en_US" data-type="horizontal" data-theme="light" data-vanity="avikarn"><a class="LI-simple-link" href='https://www.linkedin.com/in/avikarn?trk=profile-badge'>Avi Karn, PhD</a></div>

 
<script type="text/javascript" src="https://platform.linkedin.com/badges/js/profile.js" async defer></script>

<p class="about-text">
<span class="fa fa-briefcase about-icon"></span>
  Hi! My name is Avinash Karn, I go by <strong>Avi</strong>. I am very passionate about plants and cutting-edge technology. I have in-depth knowledge and a rigorous hands on experience in the field of plant breeding, quantitative genetics, computational biology and high-throughput phenotyping for over 10 years now, which I earned during my research work with plant breeders/geneticists, bioinformaticists, pathologist in my undergraduate, graduate and my postdoctoral career here in the United States. I am a creative project leader and cross-functional collaborator with an ability to excel in multifaceted fast-paced teams as evident from collaborations with different breeding teams 5 published manuscripts in high impact scientific journals and 9 in preparation, 10+ oral presentations. I intend to use my education and training in pursuing a career in an organization that is involved to improve human nutrition, health and environment for both developed and developing communities. Finally, Some of my hobbies include astronomy, painting, stamp collecting, online blogging/writing, photography and cooking! ðŸ˜‰
</p>

<hr>
<center>Skills & interests</center>
</hr>

- Skills: R, Python, Statistics, Machine Learning, Genetics, Genomics, Plant Breeding, Molecular Biology, Quantitative Genetics, Data Processing in Shell, Bioinformatics, UNIX/LINUX, BASH, HPC cluster, PERL, CSS, HTML, SQL, and GitHub
-	Lab Skills: DNA extraction, PCR, Gel Electrophoresis, Microscopy
-	Farming Skills: Plowing Field with Tractor, Seed Planting, Pollination
-	Language Skills: Fluent in English, Nepali, Hindi, Maithali, Bhojpuri, and Basic Spanish skills
-	Interests: Astronomy, Painting (watercolor), Cooking (American, Indian, Spanish), Robotics, Bird watching, Fish keeping, Gardening

<hr>
<center> <h2> Publications </h2> </center> 
<hr>

<a href="https://www.nature.com/articles/s41467-019-14280-1"> A rhAmpSeq haplotype strategy targeting the collinear core genome improves marker transferability across a diverse genus</a>
Cheng Zou, __Avinash Karn__, Bruce Reisch, Allen Nguyen, Yongming Sun, Yun Bao, Michael S. Campbell, Deanna Church, Stephen Williams, Timothy P. L. Smith, Anne Fennell, Doreen Ware, Jason Londo, Qi Sun and Lance Cadle-Davidson
(2020) *Nature Communications*

<a href="https://www.frontiersin.org/articles/10.3389/fpls.2019.00599/full">Computational analysis of AmpSeq data for targeted, high-throughput genotyping of amplicons</a>
Jonathan Fresnedo-RamÃ­rez , Shanshan Yang , Qi Sun, **Avinash Karn** , Bruce I. Reisch, Lance Cadle-Davidson (2019)
*Frontiers in Plant Sciences*

<a href="https://link.springer.com/chapter/10.1007/978-3-030-20728-1_6">Genome Wide Association Study (GWAS) on Disease Resistance in Maize. </a> Shrestha, Vivek, Mani Awale, and **Avinash Karn**. *In Disease Resistance in Crop Plants* Springer, Cham, 2019.

<a href="https://www.g3journal.org/content/7/4/1157.abstract"> Genetic Analysis of Teosinte Alleles for Kernel Composition Traits in Maize</a> **Karn, Avinash**, Gillman, J. D., & Flint-Garcia, S. A. (2017) 
*G3: Genes, Genomes, Genetics*

<a href="https://link.springer.com/article/10.1007/s11746-016-2916-4"> Development of Rigorous Fatty Acid Near-Infrared Spectroscopy Quantitation Methods in Support of Soybean Oil Improvement.</a>
**Karn, Avinash**, Heim, C., Flint-Garcia, S., Bilyeu, K., & Gillman, J. (2016)
*Journal of the American Oil Chemists*

<a href="https://www.ars.usda.gov/research/publications/publication/?seqNo115=263955">Breeding Specialty Starch Maize Using Exotic Genetic Resources for Gene Discovery of Novel Alleles and Modifiers with Materials Generated from the USDA-ARS GEM Project</a>  (2011)
**Karn, Avinash**, Brent Buckner, Diane Janick-Buckner, Mark Campbell, Michael Blanco, Marvin Scott, and Adrienne Moran Lauter .
*Maize Genetics Conference Abstracts*

<hr>
<center> <h2> Education </h2> </center> 
<hr>
#### 2012â€“2017 PhD - Plant Breeding, Genetics & Genomics â€” University of Missouri Columbia, USA
<li>Dissertation <a href="https://mospace.umsystem.edu/xmlui/handle/10355/61937">Evaluation of Teosinte Genetic Diversity for Agronomic and Domestication Traits in Maize</a> </li>
<li>Received Dr. Marcus Zuber endowment fellowship</li>
<li>Presented research at Maize genetics conference, Plant Animal Genetics conference and The Tri-Societies meeting (ASA, CSSA, SSSA)     International meetings</li>
<li>Recipient of 2017 Research in Plant Biology Award (Monetary award $750) </li>
<li>2017 MU Life Sciences Week Research Recognition award (Monetary award $275) </li>
<li>Recipient of DPS and IPG travel awards</li>

#### 2008â€“2012 BSc Agricultural Sciences â€” Truman State University, USA
<li>Received Presidentâ€™s Honorary Scholarship and Dr. Max E. Bell Scholarship</li>
<li>Presented research at student research conferences at Truman, Adair County extension center and annual Maize genetics conference</li>

<center> <h2> Work Experience </h2> </center>
<hr>

#### 2018â€“Present Computational biologist (Postdoc) â€” Cornell University, Cornell AgriTech, Geneva, NY

<ul>
  
<li>Analyzed large-scale genotyping data from Genotyping-by-Sequencing, Amplicon Sequencing, Whole genome sequencing for SNP and haplotype calling, genome alignment, genome assembly and genomic selection and prediction. </li>

<li>Co-Lead the development of a rhAmpSeq (RNase H2 enzyme-dependent amplicon sequencing) haplotype strategy targeting the collinear core genome to improve marker transferability across a diverse genus.</li>

<li>Worked with nation-wide grape breeders to apply amplicon sequencing (AmpSeq) marker data for accelerated marker-assisted breeding.</li>

<li>Built genetic maps (in Lep-Map3) of various Vitis populations, and performed quality control (QC) analysis, interval mapping and genome-wide association (GLM, MLM, FarmCPU). Build machine learning model for genomic prediction of quantitative traits</li>

<li>Managed grapevine breeding database www.VitisBase.org in collaboration with Boyce Thompson Institute (BTI) and Breeding Insight project (USDA); and led setting up trials and high-throughput phenotype data collection on FieldBook android application.</li>

<li>Big data visualization </li>

</ul>

#### 2017 Mayâ€“2017 Dec Quantitative Geneticist (Postdoc) â€” University of Illinois - Urbana-Champaign, IL

Led and managed undergraduate students and technicians in the installation of over 2,000 `minirhizotron` tubes in field to image and study maize roots with `Bartz imaging systems`; and collections of minirhizotron images, moisture data, flowering, above ground biomass and yield harvest. Identified genetics basis for maize root structure and control of plant nutrient relations in Illinois Protein Strain Recombinant Inbred (IPSRI) lines of maize

#### 2012â€”2017 Graduate Research Assistant (PhD) - University of Missouri Columbia, USA

Involved in the development of novel genetic resource and mapping populations such as near isogenic and synthetic lines derived from domesticated corn and its wild relative, Teosinte. Conducted `Joint linkage QTL analysis` and Genome wide association study to elucidate several aspects of genetic diversity in maize; genetic basis of various agronomic, fitness and yield related traits such as seed quality and composition in trials conducted with multiple replications in different environment and year. Investigated gene by gene epistatic interactions for various adaptation and domestication related traits in trials with multiple replications and environment.

#### 2009-2012 Undergraduate Research Assistant (BSc) - Truman State University, USA
Conducted molecular genetic analysis to exploit the biodiversity of exotic inbred lines from Central and South America to develop high amylose corn using SSR markers and gel electrophoresis. Assisted in preparing summer maize nurseries such as plowing field with a tractor, planting seeds with hand and dumping seed in seed planter, pollinating and collecting various agronomic data from field trials, harvesting and analyzing data. Developed gene specific molecular markers to select alleles responsible for high amylose in maize kernels with better yield, which I validated their stability in trials conducted in multiple replications in different environments. Wrote annual USDA-ARS Germplasm Enhancement of Maize (GEM) report and presented at ASTA meeting in Chicago, Illinois. 

#### 2008-2012 Cafeteria student worker - Sodexo Inc. , USA
Served and stocked food as well as performed all the cleaning related activities in the cafeteria.

#### 2011-2012 Comparative genomics researcher - Truman State University, USA
Analyzed retained homeologous paralog pairs from syntenic regions in Maize and Sorghum. Visualized genomic data of approximately 1Mb in size in `CIRCOS`, and conducted comparative genetic study in online tools such as `GeVo function in CoGe`, Annotated genes using `InterProScan` and expression analysis data from `Plexdb.org`

#### Summer 2009 & 2012 NSF TruScholars Research Program - Truman State University, USA
Designed gene specific molecular marker for a novel of starch branching enzyme Ia (SBE-1a) allele involved in high amylose starch in maize. Discovered novel recessive starch branching enzyme 1a (sbe1a) likely plays a significant role in the presence of the ae allele in elevating starch amylose from ~55% to >70% in the germplasm release 'GEMS-0067'. Conducted linkage based mapping studies to identify QTLs for high amylose starch in maize using SSR markers and `QTL cartographer` software. Sequenced starch branching enzyme allele (SBE1a) from high amylose maize line, GEMS-0067; Discovered 8bp insertion site at intron 9 resulting splicing mutation, which putatively resulted in truncated protein product post-translation.

<hr>
<div id="contactme-section">
<h1 id="contact">Contact</h1>

<p>You can <a href="mailto:akarn@mail.missouri.edu?subject=Hello from AviKarn.com">email me</a> or find me <a href="https://twitter.com/avi_karn">on Twitter</a> if you want to get in touch. I love meeting new people so don't hesitate to send a message!</p>
<p>Letâ€™s connect if you want to collaborate and or exchange ideas!</p>
<p>If you'd like to set up a meeting with me, <a href="https://calendly.com/avikarn">click here to schedule a meeting</a> based on your availability.</p>

<form action="https://formspree.io/akarn@mail.missouri.edu" method="POST" class="form" id="contact-form">
  <p>You can also send me a quick message using the form below:</p>
  <div class="row">
    <div class="col-xs-6">
      <input type="email" name="_replyto" class="form-control input-lg" placeholder="Email" title="Email">
    </div>
    <div class="col-xs-6">
      <input type="text" name="name" class="form-control input-lg" placeholder="Name" title="Name">
    </div>
  </div>
  <input type="hidden" name="_subject" value="New submission from AviKarn.com">
  <textarea type="text" name="content" class="form-control input-lg" placeholder="Message" title="Message" required="required" rows="3"></textarea>
  <input type="text" name="_gotcha" style="display:none">
  <input type="hidden" name="_next" value="./aboutme?message=Your message was sent successfully, thanks!" />
  <button type="submit" class="btn btn-lg btn-primary">Submit</button>
</form>
<hr>

<!--- Google--->
<script>
     (adsbygoogle = window.adsbygoogle || []).push({
          google_ad_client: "ca-pub-5126027065024936",
          enable_page_level_ads: true
     });
</script>

<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-123359651-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'UA-123359651-1');
</script>
<script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
