---
layout: page
title: About me
meta-title: Avi Karn, Ph.D.
subtitle: Quantitative & Computational Geneticist
---

<div id="aboutme-section">

<div class="LI-profile-badge"  data-version="v1" data-size="large" data-locale="en_US" data-type="horizontal" data-theme="light" data-vanity="avikarn"><a class="LI-simple-link" href='https://www.linkedin.com/in/avikarn?trk=profile-badge'>Avi Karn, PhD</a></div>

 
<script type="text/javascript" src="https://platform.linkedin.com/badges/js/profile.js" async defer></script>

<p class="about-text">
<span class="fa fa-briefcase about-icon"></span>
Hi! My name is Avinash Karn, I go by <strong>Avi</strong>. I am very passionate about plants and cutting-edge technology. I have in-depth knowledge and a rigorous hands on experience in the field of plant breeding, quantitative genetics, computational biology and high-throughput phenotyping for over 10 years now, which I earned during my research work with plant breeders/geneticists, bioinformaticists, pathologist in my undergraduate, graduate and my postdoctoral career here in the United States.
</p>

<p class="about-text">
<span class="fa fa-heart about-icon"></span>
"Share knowledge. Its one way to immportality" - Dalai Lama. 
I strongly believe in this saying, and I intend to use my education and training in pursuing a career in an organization that is involved to improve human nutrition, health and environment for both developed and developing communities.
</p>

<p class="about-text">
<span class="fa fa-briefcase about-icon"></span>
Since 2018 Jan, I have been a Computational biologist/Molecular breeding data scientist postdoc in <a href="http://www.vitisgen2.org/"> VitisGen2 (USDA NIFA funded) project</a> at Cornell AgriTech, Cornell University in Geneva, NY
</p>
 
<p class="about-text">
<span class="fa fa-graduation-cap about-icon"></span>
<ul>
 <li> 2012–2017 PhD - Plant Breeding, Genetics & Genomics — University of Missouri Columbia, Columbia, MO, USA</li>
 <li> 2008–2012 BSc Agricultural Sciences — Truman State University, Kirksville, MO, USA </li>
</ul>
</p>

<p class="about-text">
<span class="fa fa-code about-icon"></span>
<ul>
 <li>Technical Skills: R, Python, Statistics, Machine Learning, Genetics, Genomics, Plant Breeding, Molecular Biology, Quantitative Genetics, Data Processing in Shell, Bioinformatics, UNIX/LINUX, BASH, HPC cluster, PERL, CSS, HTML, SQL, and GitHub </li>
 <li>Lab Skills: DNA extraction, PCR, Gel Electrophoresis, Microscopy </li>
 <li>Farming Skills: Plowing Field with Tractor, Seed Planting, Pollination</li>
 <li>Language Skills: Fluent in English, Nepali, Hindi, Maithali, Bhojpuri, and Basic Spanish skills </li> 
 <li>Interests: Astronomy, Painting (watercolor), Cooking (American, Indian, Spanish), Stamp Collecting, Robotics,    Bird watching, Fish keeping, Gardening </li>
</ul>
</p>

<p class="about-text">
<span class="fa fa-globe about-icon"></span>
Grew up in <i>Birgunj, Nepal</i>; moved to <i>the United States</i> to obtain my undergraduate degree and have been living here since then.
</p>


<center> <h2> Publications </h2> </center> 

<p>
<a href="https://www.nature.com/articles/s41467-019-14280-1"> A rhAmpSeq haplotype strategy targeting the collinear core genome improves marker transferability across a diverse genus</a>
Cheng Zou, __Avinash Karn__, Bruce Reisch, Allen Nguyen, Yongming Sun, Yun Bao, Michael S. Campbell, Deanna Church, Stephen Williams, Timothy P. L. Smith, Anne Fennell, Doreen Ware, Jason Londo, Qi Sun and Lance Cadle-Davidson
(2020) *Nature Communications*
</p>

<p>
<a href="https://www.frontiersin.org/articles/10.3389/fpls.2019.00599/full">Computational analysis of AmpSeq data for targeted, high-throughput genotyping of amplicons</a>
Jonathan Fresnedo-Ramírez , Shanshan Yang , Qi Sun, **Avinash Karn** , Bruce I. Reisch, Lance Cadle-Davidson (2019)
*Frontiers in Plant Sciences*
</p>

<p>
<a href="https://link.springer.com/chapter/10.1007/978-3-030-20728-1_6">Genome Wide Association Study (GWAS) on Disease Resistance in Maize. </a> Shrestha, Vivek, Mani Awale, and **Avinash Karn**. *In Disease Resistance in Crop Plants* Springer, Cham, 2019.
</p>

<p>
<a href="https://www.g3journal.org/content/7/4/1157.abstract"> Genetic Analysis of Teosinte Alleles for Kernel Composition Traits in Maize</a> **Karn, Avinash**, Gillman, J. D., & Flint-Garcia, S. A. (2017) 
*G3: Genes, Genomes, Genetics*
</p>

<p>
<a href="https://link.springer.com/article/10.1007/s11746-016-2916-4"> Development of Rigorous Fatty Acid Near-Infrared Spectroscopy Quantitation Methods in Support of Soybean Oil Improvement.</a>
**Karn, Avinash**, Heim, C., Flint-Garcia, S., Bilyeu, K., & Gillman, J. (2016)
*Journal of the American Oil Chemists*
</p>

<p>
<a href="https://www.ars.usda.gov/research/publications/publication/?seqNo115=263955">Breeding Specialty Starch Maize Using Exotic Genetic Resources for Gene Discovery of Novel Alleles and Modifiers with Materials Generated from the USDA-ARS GEM Project</a>  (2011)
**Karn, Avinash**, Brent Buckner, Diane Janick-Buckner, Mark Campbell, Michael Blanco, Marvin Scott, and Adrienne Moran Lauter .
*Maize Genetics Conference Abstracts*
</p>

<center> <h2> Work Experience </h2> </center>


<h3>2018–Present Computational biologist (Postdoc) VitisGen2 project — Cornell University, Cornell AgriTech, Geneva, NY </h3>

 - Analyzed large-scale genotyping data from Genotyping-by-Sequencing, Amplicon Sequencing, Whole genome sequencing for SNP and haplotype calling, genome alignment, genome assembly and genomic selection and prediction

- Co-Lead the development of a rhAmpSeq (RNase H2 enzyme-dependent amplicon sequencing) haplotype strategy targeting the collinear core genome to improve marker transferability across a diverse genus

- Worked with nation-wide grape breeders to apply amplicon sequencing (AmpSeq) marker data for accelerated marker-assisted breeding

- Built genetic maps (in Lep-Map3) of various Vitis populations, and performed quality control (QC) analysis, interval mapping and genome-wide association (GLM, MLM, FarmCPU). Build machine learning model for genomic prediction of quantitative traits

- Managed grapevine breeding database www.VitisBase.org in collaboration with Boyce Thompson Institute (BTI) and Breeding Insight project (USDA); and led setting up trials and high-throughput phenotype data collection on FieldBook android application

- Big data visualization in R, Python, PERL

<h3> 2017 May–2017 Dec Quantitative Geneticist (Postdoc) — University of Illinois - Urbana-Champaign, IL </h3>

- Led the genetics basis of maize roots project in Illinois Protein Strain Recombinant Inbred (IPSRI) lines of maize; the objective of this project was to elucidate influence of overall maize root genetic architecture by evaluating them in two environments with and without application of nitrogen
- Managed the installation of over 2,000 minirhizotron tubes in maize field to image and quantify root traits, and perform genome wide association study

<h3>  2012—2017 Graduate Research Assistant (PhD) - University of Missouri Columbia, USA </h3>
<p>
Involved in the development of novel genetic resource and mapping populations such as near isogenic and synthetic lines derived from domesticated corn and its wild relative, Teosinte. Conducted `Joint linkage QTL analysis` and Genome wide association study to elucidate several aspects of genetic diversity in maize; genetic basis of various agronomic, fitness and yield related traits such as seed quality and composition in trials conducted with multiple replications in different environment and year. Investigated gene by gene epistatic interactions for various adaptation and domestication related traits in trials with multiple replications and environment.
</p>

<h3>  2009-2012 Undergraduate Research Assistant (BSc) - Truman State University, USA </h3>
<p>
Conducted molecular genetic analysis to exploit the biodiversity of exotic inbred lines from Central and South America to develop high amylose corn using SSR markers and gel electrophoresis. Assisted in preparing summer maize nurseries such as plowing field with a tractor, planting seeds with hand and dumping seed in seed planter, pollinating and collecting various agronomic data from field trials, harvesting and analyzing data. Developed gene specific molecular markers to select alleles responsible for high amylose in maize kernels with better yield, which I validated their stability in trials conducted in multiple replications in different environments. Wrote annual USDA-ARS Germplasm Enhancement of Maize (GEM) report and presented at ASTA meeting in Chicago, Illinois. 
</p>

<h3>  2008-2012 Cafeteria student worker - Sodexo Inc. , USA </h3>
<p>
Served and stocked food as well as performed all the cleaning related activities in the cafeteria.
</p>

<h3>  2011-2012 Comparative genomics researcher - Truman State University, USA </h3>
<p>
Bioinformatic analysis of the intra-genomic evolution of retained homeologous paralogs in maize. This project attempted to create a model undergraduate genome annotation community. An examination of approximately eight million base pairs of the maize genome was conducted using bioinformatic tools including: genome browsers (MaizeGDB and MaizeSequence.org), gene expression databases (PlexDB), protein functional prediction software (InterProScan. FGENESH) and comparative genome platforms (CoGe). Also, the consistency of the functional annotations with the balanced gene drive model and exhibition of subgenome bias in the retention/decay, gene density, and expression patterns of homeologous paralog pairs were also evaluated in the study.
</p>

<h3>  Summer 2009 & 2012 NSF TruScholars Research Program - Truman State University, USA </h3>
<p>
Designed gene specific molecular marker for a novel of starch branching enzyme Ia (SBE-1a) allele involved in high amylose starch in maize. Discovered novel recessive starch branching enzyme 1a (sbe1a) likely plays a significant role in the presence of the ae allele in elevating starch amylose from ~55% to >70% in the germplasm release 'GEMS-0067'. Conducted linkage based mapping studies to identify QTLs for high amylose starch in maize using SSR markers and `QTL cartographer` software. Sequenced starch branching enzyme allele (SBE1a) from high amylose maize line, GEMS-0067; Discovered 8bp insertion site at intron 9 resulting splicing mutation, which putatively resulted in truncated protein product post-translation.
</p>

</div>

<!--- Google--->
<script>
     (adsbygoogle = window.adsbygoogle || []).push({
          google_ad_client: "ca-pub-5126027065024936",
          enable_page_level_ads: true
     });
</script>

<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-123359651-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'UA-123359651-1');
</script>
<script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
